### PR TITLE
[WebDAV] Enhance file listing performance

### PR DESCRIPTION
**Pull request by BlackbitNeueMedien**
Almost every customer who we tell about Pimcore's WebDAV interface is first impressed but then says it is unusable slow. This PR is a try to improve that.

To check permissions I copied the logic from https://github.com/pimcore/pimcore/blob/4c85c70e172bffef38a14c906013463cd32d6590/bundles/AdminBundle/Controller/Admin/Asset/AssetController.php#L256-L271

### CODE REVIEW

File: models/Asset/WebDAV/Folder.php [hunk 1]
```diff
@@ -50,39 +51,56 @@ public function getChildren()
...
      */
     public function getChild($name)
     {
-        $nameParts = explode('/', $name);
-        $name = Element\Service::getValidKey($nameParts[count($nameParts) - 1], 'asset');
         $asset = null;
 
         if (is_string($name)) {
+            $nameParts = explode('/', $name);
+            $name = Element\Service::getValidKey($nameParts[count($nameParts) - 1], 'asset');
```
Comment (aarongerig) [inline models/Asset/WebDAV/Folder.php:100]:
```suggestion
            $name = Element\Service::getValidKey(substr($name, strrpos($name, '/') + 1), 'asset');
```

File: models/Asset/WebDAV/Folder.php [hunk 2]
```diff
@@ -50,39 +51,56 @@ public function getChildren()
...
-            foreach ($this->asset->getChildren() as $child) {
-                if ($child->isAllowed('view')) {
-                    try {
-                        if ($child = $this->getChild($child)) {
-                            $children[] = $child;
-                        }
-                    } catch (\Exception $e) {
-                        Logger::warning($e);
-                    }
+        $db = Db::get();
```
Comment (aarongerig) [inline models/Asset/WebDAV/Folder.php:54]: Use DI here:

```php
Pimcore\Db\ConnectionInterface $connection
```

Comment (BlackbitNeueMedien) [reply models/Asset/WebDAV/Folder.php:54]: This is not a symfony service but it gets instantiated in https://github.com/pimcore/pimcore/blob/0c8b8e9ddc97dd5107a0a4f082857da2e651151c/bundles/AdminBundle/Controller/Admin/Asset/AssetController.php#L896

Of course we could change that but as there are a lot of other classes where `Db::get()` is used (like the above mentioned `AssetController`), I do not see a problem that this PR also does.

Comment (brusch) [reply models/Asset/WebDAV/Folder.php:54]: I think it's fine for now.

**Comment by brusch** [comment]
@BlackbitNeueMedien but maybe we can avoid copy&paste and put this piece of code into a generic private function?

File: models/Asset/WebDAV/Folder.php (hunk 2 above)
Comment (BlackbitNeueMedien) [reply models/Asset/WebDAV/Folder.php]: Never mind, now I provide the parameter separately for `setCondition()`, then we do not have to manually quote it.

**Comment by BlackbitNeueMedien** [comment]
@brusch you mean the code from `AssetController`? A private method would not help here because we could not access this from both places. Possible would be a trait or to add a method in `Asset\Folder` like `getChildrenForUser()`. Or even integrate it into `getChildren()`?

**Comment by brusch** [comment]
@BlackbitNeueMedien oh, sorry, there was a misunderstanding on my side 😊 
My favorite is also `filterAccessibleByUser()` tagged as `@internal` on `Asset\Listing`, can you update your PR accordingly? 
Thanks a lot!

**Comment by BlackbitNeueMedien** [comment]
@brusch I extracted `Asset\Listing::filterAccessibleByUser()`.

**Comment by brusch** [comment]
@BlackbitNeueMedien very well done, thanks a lot! 🙏

**Pull request merged**